stdlib/buffer.gr

```diff
@@ -0,0 +1,742 @@
+/**
+ * @module Buffer: Utilities for working with buffers.
+ *
+ * @example import Buffer from "buffer"
+ *
+ * @since v0.4.0
+ */
+
+import Memory from "runtime/unsafe/memory"
+import WasmI32 from "runtime/unsafe/wasmi32"
+import Conv from "runtime/unsafe/conv"
+import Exception from "runtime/exception"
+import Int32 from "int32"
+import Bytes from "bytes"
+import String from "string"
+import { coerceNumberToWasmI32 } from "runtime/numbers"
+
+/**
+ * Buffers are an automatically growing data structure useful for storing and operating on a unknown number of bytes.
+ */
+record Buffer {
+  mut len: Number,
+  initialSize: Number,
+  mut data: Bytes,
+}
+
+let _SIZE_OFFSET = 4n;
+
+let _VALUE_OFFSET = 8n;
+
+let _8BIT_LEN = 1
+
+let _16BIT_LEN = 2
+
+let _32BIT_LEN = 4
+
+let _64BIT_LEN = 8
+
+/** Gets the size of a Bytes via its ptr */
+@disableGC
+let getSize = (ptr) => WasmI32.load(ptr, _SIZE_OFFSET)
+
+/** Doubles the size of buffer's underlying byte sequence, if the given size is larger than the size of a buffer's underlying byte sequence */
+let autogrow = (len, buf) => {
+  while (buf.len + len > Bytes.length(buf.data)) {
+    let mut n = Bytes.length(buf.data)
+    if (n == 0) n = 4 // Make sure bytes of 0 length grow too
+    buf.data = Bytes.resize(0, n, buf.data)
+  }
+}
+
+/** Gets the pointer for a char's bytes following the char's tag */
+@disableGC
+let rec getCharAsWasmI32 = (char) => {
+  let c = WasmI32.fromGrain(char)
+  WasmI32.load(c, 4n)
+}
+
+/** Gets the UTF-8 byte length of a char */
+@disableGC
+let rec getCharByteLength = (byte) => {
+  let (+) = WasmI32.add
+  let (&) = WasmI32.and
+  let (==) = WasmI32.eq
+  if ((byte & 0x80n) == 0x00n) {
+    1n
+  } else if ((byte & 0xF0n) == 0xF0n) {
+    4n
+  } else if ((byte & 0xE0n) == 0xE0n) {
+    3n
+  } else {
+    2n
+  }
+}
+
+/** Memcopies bytes from a source byte sequence to a destination byte sequence via pointers */
+@disableGC
+let rec appendBytes = (srcOff, dstOff, len, src, dst) => {
+  let (+) = WasmI32.add
+  Memory.copy(dst + _VALUE_OFFSET + dstOff, src + _VALUE_OFFSET + srcOff, len)
+}
+
+/** Throws if a given index + len is not within the bounds of a buffer's current length */
+let checkIndexIsInBounds = (i, len, buf) => {
+  if (i < 0) throw Exception.IndexOutOfBounds
+  if (i >= buf.len || i + len > buf.len) throw Exception.IndexOutOfBounds
+}
+
+/**
+ * @section Values: Functions for working with the Buffer data type.
+ */
+
+/**
+ * Gets a signed 8-bit integer starting at the given byte index.
+ * 
+ * @param index: The byte index to access
+ * @param buffer: The buffer to access
+ * @returns A 32-bit integer representing a signed 8-bit integer that starts at the given index
+ * 
+ * @since v0.4.0
+ */
+export let getInt8S = (i, b) => {
+  checkIndexIsInBounds(i, _8BIT_LEN, b)
+  Bytes.getInt8S(i, b.data)
+}
```

Comment:
Graindoc needs to match up the names you document with the parameters of your functions, so you need to update the signatures and implementations throughout this file.
```suggestion
export let getInt8S = (index, buffer) => {
  checkIndexIsInBounds(index, _8BIT_LEN, buffer)
  Bytes.getInt8S(index, buffer.data)
}
```
You should try running the `grain doc` command on this to catch this sort of thing 🔥